src/main/java/org/ice4j/util/PacketQueue.java

```diff
@@ -0,0 +1,524 @@
+        {
+            logger.warning("No handler set, the reading thread will be stopped.");
+            return;
+        }
+
+        while (!closed)
+        {
+            T pkt;
+
+            synchronized (queue)
+            {
+                pkt = queue.poll();
+                if (pkt == null)
+                {
+                    try
+                    {
+                        queue.wait(100);
+                    }
+                    catch (InterruptedException ie)
+                    {
+                    }
+                    continue;
+                }
+            }
+
+            if (queueStatistics != null)
+            {
+                queueStatistics.remove(System.currentTimeMillis());
+            }
+
+            try
+            {
+                handler.handlePacket(pkt);
+            }
+            catch (Exception e)
+            {
+                logger.warning("Failed to handle packet: " + e);
+            }
+        }
+    }
+
+    /**
+     * Adds a packet represented by a {@code byte[]} with a corresponding
+     * offset and length to this queue.
+     * @param buf the {@code byte[]} to add.
+     * @param off the offset into {@code byte[]} where data begins.
+     * @param len the length of the data.
+     */
+    public void add(byte[] buf, int off, int len)
+    {
+        add(buf, off, len, null);
+    }
+
+    /**
+     * Adds a packet represented by a {@code byte[]} with a corresponding
+     * offset and length, and a context object to this queue.
+     * @param buf the {@code byte[]} to add.
+     * @param off the offset into {@code byte[]} where data begins.
+     * @param len the length of the data.
+     * @param context an object which will be added to the queue as part of the
+     * packet.
+     */
+    public void add(byte[] buf, int off, int len, Object context)
+    {
+        if (copy)
+        {
+            byte[] newBuf = getByteArray(len);
+            System.arraycopy(buf, off, newBuf, 0, len);
+            doAdd(createPacket(newBuf, 0, len, context));
+        }
+        else
+        {
+            doAdd(createPacket(buf, off, len, context));
+        }
+    }
+
+    /**
+     * Adds a specific packet ({@code T}) instance to the queue.
+     * @param pkt the packet to add.
+     */
+    public void add(T pkt)
+    {
+        if (copy)
+        {
+            // create a new instance
+            add(getBuffer(pkt), getOffset(pkt), getLength(pkt), getContext(pkt));
+        }
+        else
+        {
+            doAdd(pkt);
+        }
+    }
+
+    /**
+     * Get an unused {@link byte[]} instance with length at least {@code len}.
+     * @param len the minimum length of the returned instance.
+     * @return a {@link byte[]} instance with length at least {@code len}.
+     */
+    private byte[] getByteArray(int len)
+    {
+        if (cache != null)
+        {
+            byte[] buf = cache.poll();
+            if (buf != null && buf.length >= len)
```

Comment:
The rationale is that if the cache contains a byte[] with insufficient length, it is released rather than returned to the cache because we are likely to run into the same situation later on.
Anyway, I have removed the whole cache thing, because it isn't required right now, and the implementation wasn't complete.